bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/values/XMLItem.java

```diff
@@ -346,14 +364,59 @@ public void addChildren(BXML seq) {
                     && appendingList.get(0).getNodeType() == TEXT) {
                 mergeAdjoiningTextNodesIntoList(leftList, appendingList);
             } else {
+                for (BXML bxml : appendingList) {
+                    addParent(bxml, this);
+                }
                 leftList.addAll(appendingList);
             }
         } else {
+            addParent(seq, this);
             leftList.add(seq);
         }
         this.children = new XMLSequence(leftList);
     }
 
+
+    private void addParent(BXML child) {
+        ensureAcyclicGraph(child, this);
+        addParent(child, this);
+    }
+
+    // This method does not ensure acyclicness of tree after adding the children. Hence this method shold only be
+    // use in scenarios where cyclic xml construction is impossible, that is only when constructing xml tree from
+    // xml literal syntax, or after ensuring the new xml tree is not cyclic.
+    private void addParent(BXML child, XMLItem thisElem) {
+        if (child.getNodeType() == ELEMENT) {
+            ((XMLItem) child).probableParents.add(new WeakReference<>(thisElem));
+        }
+    }
+
+    private void ensureAcyclicGraph(BXML newSubTree, XMLItem current) {
+        IdentityHashMap<XMLItem, XMLItem> lineageSet = new IdentityHashMap<>();
+        lineageSet.put(this, this);
+        ensureAcyclicParentGraph(newSubTree, current, lineageSet);
+    }
+
+    private void ensureAcyclicParentGraph(BXML newSubTree, XMLItem current,
+                                          Map<XMLItem, XMLItem> lineageSet) {
+        for (WeakReference<XMLItem> probableParentRef : current.probableParents) {
+            XMLItem parent = probableParentRef.get();
+            // probable parent is the actual parent.
+            if (parent.children.children.contains(current)) {
+                // If new subtree is in the lineage of current node, adding this newSubTree forms a cycle.
+                if (parent == newSubTree) {
+                    throw createXMLCycleError();
+                }
+                lineageSet.put(parent, parent);
```

Comment:
We only seem to be putting to this `lineageSet`?

Reply:
Thanks for pointing this out, this was bits remaining from a previous attempt.